README.md
<!-- SHADOW_SECTION_LOGO_START -->

<div><img alt="Logo" src="https://raw.githubusercontent.com/wessberg/helpertypes/master/documentation/asset/logo.png" height="150"   /></div>

<!-- SHADOW_SECTION_LOGO_END -->

<!-- SHADOW_SECTION_DESCRIPTION_SHORT_START -->

> A collection of TypeScript helper types

<!-- SHADOW_SECTION_DESCRIPTION_SHORT_END -->

<!-- SHADOW_SECTION_BADGES_START -->

<a href="https://npmcharts.com/compare/helpertypes?minimal=true"><img alt="Downloads per month" src="https://img.shields.io/npm/dm/helpertypes.svg"    /></a>
<a href="https://www.npmjs.com/package/helpertypes"><img alt="NPM version" src="https://badge.fury.io/js/helpertypes.svg"    /></a>
<a href="https://david-dm.org/wessberg/helpertypes"><img alt="Dependencies" src="https://img.shields.io/david/wessberg%2Fhelpertypes.svg"    /></a>
<a href="https://github.com/wessberg/helpertypes/graphs/contributors"><img alt="Contributors" src="https://img.shields.io/github/contributors/wessberg%2Fhelpertypes.svg"    /></a>
<a href="https://github.com/prettier/prettier"><img alt="code style: prettier" src="https://img.shields.io/badge/code_style-prettier-ff69b4.svg"    /></a>
<a href="https://opensource.org/licenses/MIT"><img alt="License: MIT" src="https://img.shields.io/badge/License-MIT-yellow.svg"    /></a>
<a href="https://www.patreon.com/bePatron?u=11315442"><img alt="Support on Patreon" src="https://img.shields.io/badge/patreon-donate-green.svg"    /></a>

<!-- SHADOW_SECTION_BADGES_END -->

<!-- SHADOW_SECTION_DESCRIPTION_LONG_START -->

## Description

<!-- SHADOW_SECTION_DESCRIPTION_LONG_END -->

This is a simple collection of general purpose helper types for TypeScript that can be used across a wide variety of projects.

<!-- SHADOW_SECTION_FEATURES_START -->

### Features

<!-- SHADOW_SECTION_FEATURES_END -->

- General-purpose
- Small
- Just a few types, nothing else

<!-- SHADOW_SECTION_FEATURE_IMAGE_START -->

<!-- SHADOW_SECTION_FEATURE_IMAGE_END -->

<!-- SHADOW_SECTION_BACKERS_START -->

## Backers

| <a href="https://usebubbles.com"><img alt="Bubbles" src="https://uploads-ssl.webflow.com/5d682047c28b217055606673/5e5360be16879c1d0dca6514_icon-thin-128x128%402x.png" height="70"   /></a> | <a href="https://github.com/cblanc"><img alt="Christopher Blanchard" src="https://avatars0.githubusercontent.com/u/2160685?s=400&v=4" height="70"   /></a> | <a href="https://github.com/ideal-postcodes"><img alt="Ideal Postcodes" src="https://avatars.githubusercontent.com/u/4996310?s=200&v=4" height="70"   /></a> | <a href="https://www.xerox.com"><img alt="Xerox" src="https://avatars.githubusercontent.com/u/9158512?s=200&v=4" height="70"   /></a> |
| ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------- |
| [Bubbles](https://usebubbles.com)<br><strong>Twitter</strong>: [@usebubbles](https://twitter.com/usebubbles)                                                                                | [Christopher Blanchard](https://github.com/cblanc)                                                                                                         | [Ideal Postcodes](https://github.com/ideal-postcodes)                                                                                                        | [Xerox](https://www.xerox.com)                                                                                                        |

### Patreon

<a href="https://www.patreon.com/bePatron?u=11315442"><img alt="Patrons on Patreon" src="https://img.shields.io/endpoint.svg?url=https%3A%2F%2Fshieldsio-patreon.vercel.app%2Fapi%3Fusername%3Dwessberg%26type%3Dpatrons"  width="200"  /></a>

<!-- SHADOW_SECTION_BACKERS_END -->

<!-- SHADOW_SECTION_TOC_START -->

## Table of Contents

- [Description](#description)
  - [Features](#features)
- [Backers](#backers)
  - [Patreon](#patreon)
- [Table of Contents](#table-of-contents)
- [Install](#install)
  - [npm](#npm)
  - [Yarn](#yarn)
  - [pnpm](#pnpm)
- [Usage](#usage)
- [Contributing](#contributing)
- [Maintainers](#maintainers)
- [FAQ](#faq)
  - [Will this library grow over time?](#will-this-library-grow-over-time)
- [License](#license)

<!-- SHADOW_SECTION_TOC_END -->

<!-- SHADOW_SECTION_INSTALL_START -->

## Install

### npm

```
$ npm install helpertypes
```

### Yarn

```
$ yarn add helpertypes
```

### pnpm

```
$ pnpm add helpertypes
```

<!-- SHADOW_SECTION_INSTALL_END -->

<!-- SHADOW_SECTION_USAGE_START -->

## Usage

<!-- SHADOW_SECTION_USAGE_END -->

Simply import the types you want from `helpertypes`:

```ts
import {ElementOf} from "helpertypes";

const FAVORITE_FOOD = ["pizza", "burger", "lasagna"] as const;

// "pizza"|"burger"|"lasagna"
type FavoriteFood = ElementOf<typeof FAVORITE_FOOD>;
```

<!-- SHADOW_SECTION_CONTRIBUTING_START -->

## Contributing

Do you want to contribute? Awesome! Please follow [these recommendations](./CONTRIBUTING.md).

<!-- SHADOW_SECTION_CONTRIBUTING_END -->

<!-- SHADOW_SECTION_MAINTAINERS_START -->

## Maintainers

| <a href="mailto:frederikwessberg@hotmail.com"><img alt="Frederik Wessberg" src="https://avatars2.githubusercontent.com/u/20454213?s=460&v=4" height="70"   /></a>                                                                |
| -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| [Frederik Wessberg](mailto:frederikwessberg@hotmail.com)<br><strong>Twitter</strong>: [@FredWessberg](https://twitter.com/FredWessberg)<br><strong>Github</strong>: [@wessberg](https://github.com/wessberg)<br>_Lead Developer_ |

<!-- SHADOW_SECTION_MAINTAINERS_END -->

<!-- SHADOW_SECTION_FAQ_START -->

## FAQ

<!-- SHADOW_SECTION_FAQ_END -->

### Will this library grow over time?

Yes, but only for helper types that have common applicability. Feel free to submit a PR, and we'll take a look at it then.

<!-- SHADOW_SECTION_LICENSE_START -->

## License

MIT © [Frederik Wessberg](mailto:frederikwessberg@hotmail.com) ([@FredWessberg](https://twitter.com/FredWessberg)) ([Website](https://github.com/wessberg))

<!-- SHADOW_SECTION_LICENSE_END -->


CHANGELOG.md
## [0.0.19](https://github.com/wessberg/helpertypes/compare/v0.0.18...v0.0.19) (2022-11-09)


### Features

* add ObjectLookupString helper type ([4a19d76](https://github.com/wessberg/helpertypes/commit/4a19d76c009fa4d1648c8255e2ffcb7cfe5fe56a))



## [0.0.18](https://github.com/wessberg/helpertypes/compare/v0.0.17...v0.0.18) (2021-11-18)


### Bug Fixes

* replace implementation of ObjectLookupTitle with one less prone to exceeding allowed recursive calls ([171d7fe](https://github.com/wessberg/helpertypes/commit/171d7fe129f74ce32c140c438d6fc3a7bcb7ce45))



## [0.0.17](https://github.com/wessberg/helpertypes/compare/v0.0.16...v0.0.17) (2021-10-08)


### Bug Fixes

* yet another fix of RequiredDeep and PartialDeep ([f3b366d](https://github.com/wessberg/helpertypes/commit/f3b366d7e34faa5d5236b06f6f28dbc9cd3207c0))



## [0.0.16](https://github.com/wessberg/helpertypes/compare/v0.0.15...v0.0.16) (2021-10-08)


### Bug Fixes

* yet another fix of RequiredDeep and PartialDeep ([df959c6](https://github.com/wessberg/helpertypes/commit/df959c6d22bb71b0fc9599b82b1bb17a1866126c))



## [0.0.15](https://github.com/wessberg/helpertypes/compare/v0.0.14...v0.0.15) (2021-10-08)


### Bug Fixes

* fix behavior of Next type ([e918ac1](https://github.com/wessberg/helpertypes/commit/e918ac1e659729eab32af0ea893ec2082de31967))



## [0.0.14](https://github.com/wessberg/helpertypes/compare/v0.0.13...v0.0.14) (2021-10-08)


### Features

* add helper: PickIntersectingKeys ([6ff5552](https://github.com/wessberg/helpertypes/commit/6ff55521681311dc2fb20d0e298e2e623f7ab5d1))
* add new helper: RequiredDeep ([cb12d71](https://github.com/wessberg/helpertypes/commit/cb12d71de7c46932b7d2a66936b9b30cc5e60a02))



## [0.0.13](https://github.com/wessberg/helpertypes/compare/v0.0.12...v0.0.13) (2021-10-06)


### Features

* add new helper types: PickOptional and PickRequired ([1a9d2eb](https://github.com/wessberg/helpertypes/commit/1a9d2ebd96665749ebb75b48427d143f0eb4ece0))



## [0.0.12](https://github.com/wessberg/helpertypes/compare/v0.0.11...v0.0.12) (2021-09-15)


### Features

* add additional helper types ([7a6eb8e](https://github.com/wessberg/helpertypes/commit/7a6eb8ed4b9aced393a3825880f6a93dc3464c06))



## [0.0.11](https://github.com/wessberg/helpertypes/compare/v0.0.10...v0.0.11) (2021-09-06)


### Bug Fixes

* add Function to ignored types in PartialDeep ([ee8d6f2](https://github.com/wessberg/helpertypes/commit/ee8d6f24f15d81d5870ff20384d42ccc0a7aff6d))



## [0.0.10](https://github.com/wessberg/helpertypes/compare/v0.0.9...v0.0.10) (2021-09-03)


### Features

* add 'SplitRecord' ([465563b](https://github.com/wessberg/helpertypes/commit/465563bb68fc0f98884417759ef8e30d9d113d29))
* add 'SplitRecord' ([0a3fce3](https://github.com/wessberg/helpertypes/commit/0a3fce3a6dd0d7946050ff6c90d35d7cbabb1531))



## [0.0.9](https://github.com/wessberg/helpertypes/compare/v0.0.8...v0.0.9) (2021-09-03)


### Features

* add 'ArrayPiercingObjectLookupTuple' ([8ef5a03](https://github.com/wessberg/helpertypes/commit/8ef5a03552b6f446c2759eefce631b02487985e9))



## [0.0.8](https://github.com/wessberg/helpertypes/compare/v0.0.7...v0.0.8) (2021-09-02)


### Bug Fixes

* add support for readonly arrays in ObjectLookupTuple ([371a7ea](https://github.com/wessberg/helpertypes/commit/371a7ea2f414b8398ce52bd55ebd0fce00448bb7))



## [0.0.7](https://github.com/wessberg/helpertypes/compare/v0.0.6...v0.0.7) (2021-09-02)


### Features

* add 'ObjectLookupTuple' helper type ([5263180](https://github.com/wessberg/helpertypes/commit/5263180332d55d628a8b6afc4daa3dcb79941a01))



## [0.0.6](https://github.com/wessberg/helpertypes/compare/v0.0.5...v0.0.6) (2021-09-02)


### Bug Fixes

* include RegExp in as simple value in PartialDeep check ([aebe2d6](https://github.com/wessberg/helpertypes/commit/aebe2d687db6956cca27f7e9c30b0a2ee2e13a47))



## [0.0.5](https://github.com/wessberg/helpertypes/compare/v0.0.4...v0.0.5) (2021-09-02)


### Features

* expose 'Prev' and 'Next' types ([2b365f8](https://github.com/wessberg/helpertypes/commit/2b365f88e6e65c47b93e137a1b62aa41c1c7d9e1))



## [0.0.4](https://github.com/wessberg/helpertypes/compare/v0.0.3...v0.0.4) (2021-08-22)


### Features

* add additional helper type: RequiredExcept ([6b6998f](https://github.com/wessberg/helpertypes/commit/6b6998ff68cc1e6d83abb2f207478adee319042a))



## [0.0.3](https://github.com/wessberg/helpertypes/compare/v0.0.2...v0.0.3) (2021-06-06)


### Features

* add helpers: UppercaseKeys, LowercaseKeys, CapitalizeKeys, UncapitalizeKeys ([6ce3bfd](https://github.com/wessberg/helpertypes/commit/6ce3bfdc99ab438d7d1000b18f7a9a717668b821))



## [0.0.2](https://github.com/wessberg/helpertypes/compare/v0.0.1...v0.0.2) (2021-05-27)


### Features

* add helper: Mutable ([ff00233](https://github.com/wessberg/helpertypes/commit/ff0023305c121694044d58dcf50db7be228dea6d))



## 0.0.1 (2021-05-26)





src/index.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/**
 * Makes T a union of T, null, and undefined
 */
export type Nullable<T> = T | null | undefined;

/**
 * Makes all properties of T mutable
 */
export type Mutable<T> = {
	-readonly [P in keyof T]: T[P];
};

/**
 * Picks the element type of the given iterable.
 * For example, for the given array literal type: ["foo", "bar", "baz"],
 * the type is "foo"|"bar"|"baz"
 */
export type ElementOf<Iterable> = Iterable extends (infer ArrayElementType)[]
	? ArrayElementType
	: Iterable extends readonly (infer ReadonlyArrayElementType)[]
	? ReadonlyArrayElementType
	: Iterable extends Set<infer SetElementType>
	? SetElementType
	: Iterable extends string
	? Iterable
	: never;

/**
 * Something that is an array of T's or a single T
 */
export type MaybeArray<T> = T[] | T;

/**
 * Ensures that T is an array of the element type of T
 */
export type EnsureArray<T> = T extends (infer U)[] ? U[] : T[];

/**
 * A variant of Required where not only the keys are required, but also the values
 */
export type RequiredValues<T> = {
	[P in keyof T]-?: Exclude<T[P], null | undefined>;
};

/**
 * Picks only keys for which there is an optional modifier
 */
export type PickOptional<T> = Pick<
	T,
	Exclude<
		{
			[K in keyof T]: T extends Record<K, T[K]> ? never : K;
		}[keyof T],
		undefined
	>
>;

/**
 * Picks only keys for which there is no optional modifier
 */
export type PickRequired<T> = Pick<
	T,
	Exclude<
		{
			[K in keyof T]: T extends Record<K, T[K]> ? K : never;
		}[keyof T],
		undefined
	>
>;

/**
 * Picks those keys from A that intersects with those from B
 */
export type PickIntersectingKeys<A, B> = Pick<
	A,
	Exclude<
		{
			[Key in keyof A]: Key extends keyof B ? Key : never;
		}[keyof A],
		undefined
	>
>;

/**
 * Picks from T the properties for which their type matches Type
 */
export type PickMembersOfType<T, Type> = Pick<
	T,
	{
		[Key in keyof T]: T[Key] extends Type ? Key : never;
	}[keyof T]
>;

export type PickMembersOfNotOfType<T, Type> = Pick<
	T,
	{
		[Key in keyof T]: T[Key] extends Type ? never : T[Key] extends Type | undefined ? never : Key;
	}[keyof T]
>;

/**
 * Builds up a lookup path into a record via tuple elements. For example, for the record `{ a: {b: {c: string}}}`, a valid value could be `["a", "b", "c"]`
 * It takes as the second optional type argument the maximum depth it can recursively descent into the target object (default: 10).
 * When an Array is discovered, indexes into the array can be provided. For example, for the record `{a: {foo: string}[]}`, a value like `["a", 0, "foo"]` is allowed.
 */
export type ObjectLookupTuple<T, MaxDepth extends number = 10> = ObjectLookupTupleHelper<T, [], MaxDepth>;
type ObjectLookupTupleHelper<T, Acc extends PropertyKey[], MaxDepth extends number, CurrentDepth extends number = 0> = {
	[Key in keyof T]: CurrentDepth extends MaxDepth
		? [...Acc, Key]
		: T[Key] extends IgnoredLookupValue
		? [...Acc, Key]
		: T[Key] extends (infer El)[] | readonly (infer El)[]
		? [...Acc, Key] | ObjectLookupTupleHelper<El, [...Acc, Key, number], MaxDepth, Next<CurrentDepth>>
		: [...Acc, Key] | ObjectLookupTupleHelper<NonNullable<T[Key]>, [...Acc, Key, number], MaxDepth, Next<CurrentDepth>>;
}[keyof T];

/**
 * A variant of ObjectLookupTuple that will walk into arrays and allow looking up members of their records as part of the lookup path. For
 * example, for the record `{a: {foo: string}[]}`, a value like `["a", "foo"]` is allowed, even though 'foo' is part of a record within an array
 */
export type ArrayPiercingObjectLookupTuple<T, MaxDepth extends number = 10> = ArrayPiercingObjectLookupTupleHelper<T, [], MaxDepth>;
type ArrayPiercingObjectLookupTupleHelper<T, Acc extends PropertyKey[], MaxDepth extends number, CurrentDepth extends number = 0> = {
	[Key in keyof T]: CurrentDepth extends MaxDepth
		? [...Acc, Key]
		: T[Key] extends IgnoredLookupValue
		? [...Acc, Key]
		: T[Key] extends (infer El)[] | readonly (infer El)[]
		? [...Acc, Key] | ArrayPiercingObjectLookupTupleHelper<El, [...Acc, Key], MaxDepth, Next<CurrentDepth>>
		: [...Acc, Key] | ArrayPiercingObjectLookupTupleHelper<NonNullable<T[Key]>, [...Acc, Key], MaxDepth, Next<CurrentDepth>>;
}[keyof T];

/**
 * Behaves similar to ObjectLookupTuple, but builds up strings that dot into objects. For example, for the record `{ a: {b: {c: string}}}`, a valid value could be `a.b.c`
 */
export type ObjectLookupString<T, MaxDepth extends number = 10, StopAtArrays extends boolean = false> = ObjectLookupStringHelper<T, ``, MaxDepth, 0, StopAtArrays>;
type MaybeSuffixWithDot<T extends string> = T extends `` ? T : `${T}.`;
type ObjectLookupStringHelper<T, Acc extends string, MaxDepth extends number, CurrentDepth extends number = 0, StopAtArrays extends boolean = false> = {
	[Key in keyof T]: Key extends string
		? CurrentDepth extends MaxDepth
			? `${MaybeSuffixWithDot<Acc>}${Key}`
			: T[Key] extends IgnoredLookupValue
			? `${MaybeSuffixWithDot<Acc>}${Key}`
			: T[Key] extends (infer El)[] | readonly (infer El)[]
			? StopAtArrays extends true
				? `${MaybeSuffixWithDot<Acc>}${Key}`
				: `${MaybeSuffixWithDot<Acc>}${Key}.${number}` | El extends IgnoredLookupValue
				? `${MaybeSuffixWithDot<Acc>}${Key}.${number}`
				: ObjectLookupStringHelper<El, `${MaybeSuffixWithDot<Acc>}${Key}.${number}`, MaxDepth, Next<CurrentDepth>, StopAtArrays>
			: `${MaybeSuffixWithDot<Acc>}${Key}` | ObjectLookupStringHelper<NonNullable<T[Key]>, `${MaybeSuffixWithDot<Acc>}${Key}`, MaxDepth, Next<CurrentDepth>, StopAtArrays>
		: never;
}[keyof T];

/**
 * A variant of Partial that is deep.
 * This means that every value is a partial recursively
 * while still preserving primitive or built-in types as they are
 */
export type PartialDeep<T, MaxDepth extends number = 10, CurrentDepth extends number = 0> = CurrentDepth extends MaxDepth
	? T
	: T extends IgnoredLookupValue
	? T
	: T extends (infer ArrayElement)[]
	? PartialDeep<ArrayElement, MaxDepth, Next<CurrentDepth>>[]
	: T extends readonly (infer ReadonlyArrayElement)[]
	? readonly PartialDeep<ReadonlyArrayElement, MaxDepth, Next<CurrentDepth>>[]
	: T extends Iterable<infer IterableType>
	? Iterable<PartialDeep<IterableType, MaxDepth, Next<CurrentDepth>>>
	: {[P in keyof T]?: PartialDeep<T[P], MaxDepth, Next<CurrentDepth>>};

/**
 * A variant of Required that is deep.
 * This means that every key is required recursively
 * while still preserving primitive or built-in types as they are
 */
export type RequiredDeep<T, MaxDepth extends number = 10, CurrentDepth extends number = 0> = CurrentDepth extends MaxDepth
	? T
	: T extends IgnoredLookupValue
	? T
	: T extends (infer ArrayElement)[]
	? RequiredDeep<ArrayElement, MaxDepth, Next<CurrentDepth>>[]
	: T extends readonly (infer ReadonlyArrayElement)[]
	? readonly RequiredDeep<ReadonlyArrayElement, MaxDepth, Next<CurrentDepth>>[]
	: T extends Iterable<infer IterableType>
	? Iterable<RequiredDeep<IterableType, MaxDepth, Next<CurrentDepth>>>
	: {[P in keyof T]-?: RequiredDeep<T[P], MaxDepth, Next<CurrentDepth>>};

/**
 * A partial of T except for the keys given in K
 */
export type PartialExcept<T, K extends keyof T> = Omit<Partial<T>, K> & Pick<T, K>;

/**
 * A partial of T except for the keys given in K
 */
export type PartialDeepExcept<T, K extends keyof T> = Omit<PartialDeep<T>, K> & Pick<T, K>;

/**
 * Require all keys in T, except for the keys given in K
 */
export type RequiredExcept<T, K extends keyof T> = Omit<Required<T>, K> & Pick<T, K>;

/**
 * Splits a record into all of its possible property intersections.
 * For example, for the record {a: 1; b: 2>}, splitting the record will
 * get the type {a: 1} | {b: 2} back.
 */
export type SplitRecord<T, U = T> = Exclude<
	{
		[Key in keyof T]: {
			[K in Key]: K extends keyof U ? T[K] : never;
		};
	}[keyof T],
	undefined
>;

/**
 * A variant of `SplitRecord` that includes all similar keys.
 * See the documentation for `SimilarObjectKeys` for more details.
 */
export type SplitRecordWithSimilarKeys<T, U = T> = Exclude<
	{
		[Key in keyof T]: Key extends keyof U ? SimilarObjectKeys<T, Key & string> : never;
	}[keyof T],
	undefined
>;

/**
 * Selects only the properties from an object for which the keys include the substring(s) given in U.
 * For example, for the object {foo: 1, fooBar: 2, barFoo: 3, baz: 4}, a value of "foo" for U will return an object with only the keys "foo", "fooBar", and "barFoo"
 */
export type SimilarObjectKeys<T, U extends keyof T & string> = PickMembersOfNotOfType<
	{
		[TKey in keyof T]: TKey extends U
			? T[TKey]
			: TKey extends `${infer _Prefix}${Capitalize<U>}`
			? T[TKey] | undefined
			: TKey extends `${infer _Prefix}${U}`
			? T[TKey] | undefined
			: TKey extends `${U}${infer _Suffix}`
			? T[TKey] | undefined
			: never;
	},
	never
>;

/**
 * An arbitrary Function that takes any amount of arguments and returns anything
 */
export type ArbitraryFunction<ReturnType = unknown> = (...args: never[]) => ReturnType;
export type ArbitraryConstructor<T> = new (...args: never[]) => T;

/**
 * Gets the length of T
 */
export type GetLength<T extends unknown[]> = T extends {length: infer L} ? L : never;

/**
 * Gets the first element of T
 */
export type GetFirst<T extends unknown[]> = T[0];

/**
 * Gets the last element of T
 */
export type GetLast<T extends unknown[]> = T[Prev<GetLength<T>>];

/**
 * Gets the Nth parameter of the given function
 */
export type NthParameter<T extends ArbitraryFunction, Parameter extends number> = Parameters<T>[Parameter];

/**
 * Gets the first parameter of the given function
 */
export type FirstParameter<T extends ArbitraryFunction> = GetFirst<Parameters<T>>;

/**
 * Gets the last parameter of the given function
 */
export type LastParameter<T extends ArbitraryFunction> = GetLast<Parameters<T>>;

export type UncapitalizeKeys<T> = {[Key in keyof T as Uncapitalize<string & Key>]: T[Key]};
export type CapitalizeKeys<T> = {[Key in keyof T as Capitalize<string & Key>]: T[Key]};
export type LowercaseKeys<T> = {[Key in keyof T as Lowercase<string & Key>]: T[Key]};
export type UppercaseKeys<T> = {[Key in keyof T as Uppercase<string & Key>]: T[Key]};

export type PrefixObjectKeys<T, Prefix extends string, EnsureCamelCase = false> = {[Key in keyof T as SuffixKey<Key & string, Prefix, EnsureCamelCase>]: T[Key]};
export type SuffixObjectKeys<T, Suffix extends string, EnsureCamelCase = false> = {[Key in keyof T as PrefixKey<Key & string, Suffix, EnsureCamelCase>]: T[Key]};
export type PrefixKey<T extends string, Prefix extends string, EnsureCamelCase = false> = `${Prefix}${EnsureCamelCase extends true ? Capitalize<T> : T}`;
export type SuffixKey<T extends string, Suffix extends string, EnsureCamelCase = false> = `${T}${EnsureCamelCase extends true ? Capitalize<Suffix> : Suffix}`;

// Internal helpers

// Internal helpers
export type Prev<T extends number> = [
	-1,
	0,
	1,
	2,
	3,
	4,
	5,
	6,
	7,
	8,
	9,
	10,
	11,
	12,
	13,
	14,
	15,
	16,
	17,
	18,
	19,
	20,
	21,
	22,
	23,
	24,
	25,
	26,
	27,
	28,
	29,
	30,
	31,
	32,
	33,
	34,
	35,
	36,
	37,
	38,
	39,
	40,
	41,
	42,
	43,
	44,
	45,
	46,
	47,
	48,
	49,
	50,
	51,
	52,
	53,
	54,
	55,
	56,
	57,
	58,
	59,
	60,
	61,
	62,
	63,
	64,
	65,
	66,
	67,
	68,
	69,
	70,
	71,
	72,
	73,
	74,
	75,
	76,
	77,
	78,
	79,
	80,
	81,
	82,
	83,
	84,
	85,
	86,
	87,
	88,
	89,
	90,
	91,
	92,
	93,
	94,
	95,
	96,
	97,
	98,
	99,
	100
][T];

export type Next<T extends number> = [
	1,
	2,
	3,
	4,
	5,
	6,
	7,
	8,
	9,
	10,
	11,
	12,
	13,
	14,
	15,
	16,
	17,
	18,
	19,
	20,
	21,
	22,
	23,
	24,
	25,
	26,
	27,
	28,
	29,
	30,
	31,
	32,
	33,
	34,
	35,
	36,
	37,
	38,
	39,
	40,
	41,
	42,
	43,
	44,
	45,
	46,
	47,
	48,
	49,
	50,
	51,
	52,
	53,
	54,
	55,
	56,
	57,
	58,
	59,
	60,
	61,
	62,
	63,
	64,
	65,
	66,
	67,
	68,
	69,
	70,
	71,
	72,
	73,
	74,
	75,
	76,
	77,
	78,
	79,
	80,
	81,
	82,
	83,
	84,
	85,
	86,
	87,
	88,
	89,
	90,
	91,
	92,
	93,
	94,
	95,
	96,
	97,
	98,
	99,
	100,
	101
][T];

/**
 * A type indicating something that shouldn't be traversed into when mapping over objects, lists, or other similar data structures
 */
export type IgnoredLookupValue =
	| string
	| number
	| bigint
	| symbol
	| boolean
	| undefined
	| null
	| Date
	| RegExp
	| CallableFunction
	| Set<unknown>
	| WeakSet<never>
	| Map<unknown, unknown>
	| WeakMap<never, unknown>;
